<a id="item31"></a>


# Importing Libraries

In [104]:
import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error



<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [105]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 


# EDA

In [106]:
# Check size of the data
concrete_data.shape

(1030, 9)

In [107]:
# describe the data
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [108]:
# check null values
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

# Data Pre-processing

#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [109]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [110]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [111]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

# A. Build a Baseline Model

In [112]:
mseA = []
def training_model_A():
    for i in range(50):
        x_train,x_test,y_train,y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)
        
         # create model
        model = Sequential()
        model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
        model.add(Dense(1))

        # compile model
        model.compile(optimizer='adam', loss='mean_squared_error')
        
        # training model
        model.fit(x_train, y_train, epochs=50, verbose='auto')
        y_pred = model.predict(x_test)

        # Model evaluation on test sample
        result = np.sqrt(mean_squared_error(y_test,y_pred))
        mseA.append(result)
        
training_model_A()

Epoch 1/50
23/23 [==============================] - 1s 1ms/step - loss: 18774.6641
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 4624.3638
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 3625.0991
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 2908.5505
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 2328.8511
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1846.3738
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1446.5131
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1178.7750
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 967.4627
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 826.6692
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 723.4108
Epoch 12/50
23/23 [==============================] - 0s 1ms/step - loss: 643.0231
Epoch 13/50
23/2

# Mean and Standard Deviation (Part A)

In [113]:
# Mean and standard deviation for part A
print("Mean: ",np.mean(mseA))
print("Standard Deviation: ",np.std(mseA))

Mean:  16.291066192575737
Standard Deviation:  8.766012140218578


# B. Normalize the Data

In [114]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
n_cols = predictors_norm.shape[1] # number of predictors
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [115]:
mseB = []
def training_model_B():
    for i in range(50):
        x_train,x_test,y_train,y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)
        
         # create model
        model = Sequential()
        model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
        model.add(Dense(1))

        # compile model
        model.compile(optimizer='adam', loss='mean_squared_error')
        
        # training model
        model.fit(x_train, y_train, epochs=50, verbose='auto')
        y_pred = model.predict(x_test)

        # Model evaluation on test sample
        result = np.sqrt(mean_squared_error(y_test,y_pred))
        mseB.append(result)
        
training_model_B()

Epoch 1/50
23/23 [==============================] - 2s 2ms/step - loss: 1651.7921
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1635.9476
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 1620.9242
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1606.1349
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 1591.4855
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1576.6191
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1561.5895
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1546.0878
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1530.0033
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1513.2797
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1495.3982
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 1476.8652
Epoch 13/50
2

# Mean and Standard Deviation (Part B)

In [116]:
# Mean and standard deviation for part B
print("Mean: ",np.mean(mseB))
print("Standard Deviation: ",np.std(mseB))

Mean:  18.851941824870767
Standard Deviation:  1.9660661231440062


# Mean and Standard Deviation Comparison (Part A & B)
### It appear that the mean for part B is worse than part A but the standard deviation for part B is much better

In [117]:
# Mean and standard deviation for part A
print("Mean Part A: ",np.mean(mseA))
print("Standard Deviation Part A: ",np.std(mseA))

# Mean and standard deviation for part B
print("Mean Part B: ",np.mean(mseB))
print("Standard Deviation: ",np.std(mseB))

Mean Part A:  16.291066192575737
Standard Deviation Part A:  8.766012140218578
Mean Part B:  18.851941824870767
Standard Deviation:  1.9660661231440062


# Part C: Increase The Number Of Epochs

In [118]:
mseC = []
def training_model_C():
    for i in range(50):
        x_train,x_test,y_train,y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)
        
         # create model
        model = Sequential()
        model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
        model.add(Dense(1))

        # compile model
        model.compile(optimizer='adam', loss='mean_squared_error')
        
        # training model
        model.fit(x_train, y_train, epochs=100, verbose='auto')
        y_pred = model.predict(x_test)

        # Model evaluation on test sample
        result = np.sqrt(mean_squared_error(y_test,y_pred))
        mseC.append(result)
        
training_model_C()

Epoch 1/100
23/23 [==============================] - 1s 2ms/step - loss: 1588.7434
Epoch 2/100
23/23 [==============================] - 0s 2ms/step - loss: 1571.8790
Epoch 3/100
23/23 [==============================] - 0s 2ms/step - loss: 1555.6096
Epoch 4/100
23/23 [==============================] - 0s 2ms/step - loss: 1539.4248
Epoch 5/100
23/23 [==============================] - 0s 2ms/step - loss: 1523.3998
Epoch 6/100
23/23 [==============================] - 0s 2ms/step - loss: 1507.2769
Epoch 7/100
23/23 [==============================] - 0s 2ms/step - loss: 1491.0804
Epoch 8/100
23/23 [==============================] - 0s 2ms/step - loss: 1474.4171
Epoch 9/100
23/23 [==============================] - 0s 2ms/step - loss: 1457.3127
Epoch 10/100
23/23 [==============================] - 0s 2ms/step - loss: 1439.8884
Epoch 11/100
23/23 [==============================] - 0s 2ms/step - loss: 1421.5573
Epoch 12/100
23/23 [==============================] - 0s 2ms/step - loss: 1402.5126
E

# Mean and Standard Deviation Comparison (Part B & C)
### It appear that the mean and standard deviation for part C is much better than part B

In [119]:
# Mean and standard deviation for part C
print("Mean Part C: ",np.mean(mseC))
print("Standard Deviation Part C: ",np.std(mseC))

# Mean and standard deviation for part B
print("Mean Part B: ",np.mean(mseB))
print("Standard Deviation: ",np.std(mseB))

Mean Part C:  12.566649024440729
Standard Deviation Part C:  0.5174244714875521
Mean Part B:  18.851941824870767
Standard Deviation:  1.9660661231440062


# Part D: Increase The Number Of Hidden Layers

In [121]:
mseD = []
def training_model_D():
    for i in range(50):
        x_train,x_test,y_train,y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)
        
         # create model
        model = Sequential()
        model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
        model.add(Dense(10, activation='relu'))
        model.add(Dense(10, activation='relu'))
        model.add(Dense(1))

        # compile model
        model.compile(optimizer='adam', loss='mean_squared_error')
        
        # training model
        model.fit(x_train, y_train, epochs=50, verbose='auto')
        y_pred = model.predict(x_test)

        # Model evaluation on test sample
        result = np.sqrt(mean_squared_error(y_test,y_pred))
        mseD.append(result)
        
training_model_D()

Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 1597.9327
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1574.3069
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 1549.3601
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1505.3309
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 1423.0067
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1291.9100
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1110.7604
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 887.3218
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 643.7943
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 435.7431
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 300.3685
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 239.2434
Epoch 13/50
23/23 

# Mean and Standard Deviation Comparison (Part B & D)
### It appear that the mean and standard deviation for part D is much better than part B

In [122]:
# Mean and standard deviation for part D
print("Mean Part D: ",np.mean(mseD))
print("Standard Deviation Part D: ",np.std(mseD))

# Mean and standard deviation for part B
print("Mean Part B: ",np.mean(mseB))
print("Standard Deviation: ",np.std(mseB))

Mean Part D:  11.192414721207983
Standard Deviation Part D:  0.555063794168192
Mean Part B:  18.851941824870767
Standard Deviation:  1.9660661231440062
